graspologic/cluster/rclust.py

```diff
@@ -0,0 +1,332 @@
+# Copyright (c) Microsoft Corporation and contributors.
+# Licensed under the MIT License.
+
+import numpy as np
+from anytree import NodeMixin
+
+from .kclust import KMeansCluster
+from .autogmm import AutoGMMCluster
+
+from sklearn.utils.validation import check_is_fitted
+from sklearn.utils import check_array
+from sklearn.base import BaseEstimator
+
+
+def _check_common_inputs(min_components, max_components, cluster_kws, embed_kws):
+    if not isinstance(min_components, int):
+        raise TypeError("min_components must be an int")
+    elif min_components < 1:
+        raise ValueError("min_components must be > 0")
+
+    if not isinstance(max_components, int):
+        raise TypeError("max_components must be an int")
+    elif max_components < 1:
+        raise ValueError("max_components must be > 0")
+    elif max_components < min_components:
+        raise ValueError("max_components must be >= min_components")
+
+    if embed_kws is not None:
+        if not isinstance(embed_kws, dict):
+            raise TypeError("embed_kws must be a dict")
+
+    if not isinstance(cluster_kws, dict):
+        raise TypeError("cluster_kws must be a dict")
+
+
+class RecursiveCluster(NodeMixin, BaseEstimator):
+    """
+    Recursively clusters data based on a chosen clustering algorithm.
+    This algorithm implements a "divisive" or "top-down" approach.
+
+    Parameters
+    ----------
+    cluster_method : str {"gmm", "kmeans"}, defaults to "gmm".
+        The clustering method chosen to apply.
+    min_components : int, defaults to 1.
+        The minimum number of mixture components to consider for the first
+        split if "gmm" is selected as cluster_method; and is set to 1
+        for later splits.
+        If "kmeans" is selected, it is set to 2 for all splits.
+    max_components : int, defaults to 2.
+        The maximum number of mixture components or clusters to consider
+        at each split.
+    min_split : int, defaults to 1.
+        The minimum size of a cluster to be split on.
+    max_level : int, defaults to 4.
+        The maximum number of times to recursively cluster the data.
+    delta_criter : float or None, positive, defaults to None
```

Comment:
is None the same as 0?

Reply:
The way I imagine it should work is if the user wants to apply this constraint, they would specify a positive value, if no input is provided, it won't run. In the current code, 0 is not considered as a valid input. So I think right now 0 is not treated as None.

Reply:
i thought it was doing "split if score_2 > score_1 + delta_crier" or the other way around, or something close to that, so I thought 0 would be a valid input. if I'm wrong, can you tell me what it's actually doing, and what values are allowed?

Reply:
@bdpedigo I think we are talking about the same thing (this [line](https://github.com/microsoft/graspologic/blob/d5adacd2d95b3675e04924341ed088fb6596d9b0/graspologic/cluster/rclust.py#L197) is the same as your inequality). I was just thinking that if someone wants to make it harsher to accept a model of more clusters, they would input some value strictly greater than 0 (since if =0 it would be the same as what agmm does); None means don't do anything extra. But you are right 0 is not different from None here. I'll allow all non-negative values and just set 0 as default. Guess that's less confusing.

Reply:
sounds good. any scalar value seems fine then?

Reply:
There is a nice summary of the rules of thumbs for both BIC & AIC in this [book](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=fabozzi+The+Basics+of+Financial+Econometrics%3A+Tools%2C+Concepts%2C+and+Asset+Management+Applications.&btnG=) that provides the original sources (specifically, on pages 401 & 403 [here](https://onlinelibrary.wiley.com/doi/pdf/10.1002/9781118856406.app5)). For both criteria, a difference of 10 seems to be strong enough (also what I used in the test).